app/recipes/tests/test_recipes_api.py

```diff
@@ -0,0 +1,259 @@
+        self.assertEqual(created_ingredients[2].name, 'Butter')
+
+    def test_create_recipe_invalid(self):
+        """POST /recipes/ with invalid data should return 400"""
+        payload = {
+            'type': 'Horse',
+            'legs': '4',
+            'speeds': ['TROT', 'CANTER', 'GALLOP']
+        }
+
+        response = self.client.post(RECIPES_URL, payload, format='json')
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_read_recipes(self):
+        """GET /recipes/ should return all recipes"""
+        recipe_1_id = create_sample_recipe(
+            'Cheese sandwich',
+            ['Bread', 'Cheese'],
+            'Bread on the outside, cheese in the middle',
+        ).id
+        recipe_2_id = create_sample_recipe(
+            'Just an apple',
+            ['Apple']
+        ).id
+
+        response = self.client.get(RECIPES_URL)
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.json(), [
+            {
+                'id': recipe_1_id,
+                'name': 'Cheese sandwich',
+                'description': 'Bread on the outside, cheese in the middle',
+                'ingredients': [
+                    {'name': 'Bread'},
+                    {'name': 'Cheese'},
+                ]
+            },
+            {
+                'id': recipe_2_id,
+                'name': 'Just an apple',
+                'description': '',
+                'ingredients': [
+                    {'name': 'Apple'}
+                ]
+            }
+        ])
+
+    def test_read_recipes_with_name_filter(self):
+        """GET /recipes/?name={name} should return all recipes whose name
+            partially matches the name param"""
+        sandwich_1_id = create_sample_recipe(
+            'Cheese sandwich',
+            ['Bread', 'Cheese', 'Butter'],
+            'Bread on the outside, cheese in the middle',
+        ).id
+        sandwich_2_id = create_sample_recipe(
+            'Bread sandwich',
+            ['Bread']
+        ).id
+        create_sample_recipe(
+            'Garlic bread',
+            ['Bread', 'Garlic', 'Butter']
+        )
+
+        response = self.client.get(RECIPES_URL, {'name': 'sand'})
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.json(), [
+            {
+                'id': sandwich_1_id,
+                'name': 'Cheese sandwich',
+                'description': 'Bread on the outside, cheese in the middle',
+                'ingredients': [
+                    {'name': 'Bread'},
+                    {'name': 'Cheese'},
+                    {'name': 'Butter'},
+                ]
+            },
+            {
+                'id': sandwich_2_id,
+                'name': 'Bread sandwich',
+                'description': '',
+                'ingredients': [{'name': 'Bread'}]
+            },
```

Comment:
Maybe add a comment here saying the "Garlic bread" is not present. Or the technique of destructuring the payload identifiers might also help make it a bit more obvious what was filtered.